src/algebra/bit manipulation.md

```diff
@@ -0,0 +1,214 @@
+
+There are two operators for shifting bits.
+
+**>>** One bit to the right. Which would be the same as a division by a power of two.
+
+**<<** One bit to the left. And this would be a multiplication, equally by a power of two.
+
+Let's see an example.
+
+Let's say we have a number $N$, which. $N = 4$.
+
+The binary representation of **4** is:
+
+$4_2 = 100$
+
+If we shift one bit to the left, then we would be adding a bit to the end, in this case a 0.
+
+$4_2 << 1 = 100 << 1 = 1000$
+
+$1000 = 8_2$ and clearly, $4 * 2 = 8$
+
+So, for a division by two, it would be exactly the same, but using the operator $<<$.
+
+$4_2 >> 1 = 100 >> 1 = 10$
+
+In this case, we would remove one bit from the end, this is what is called the **least significant bit**, the rightmost bit of the representation.
+
+Of course, the **most significant bit** is the leftmost bit.
+
+And then, how could it carry out operations with other powers, greater than two?
+
+Easy, you just have to remember the exponent of two, to create said power. In other words:
+
+$4 = 2 ^ 2$, so to do multiplication or division operations with **4**, we must move **2 bits**.
+
+$8 = 2 ^ 3$, so for **8** it would be **3 bits**. *And so on*.	
+
+
+* ## Bitwise operators.
+
+ 	* **&** : The bitwise AND operator compares each bit of its first operand with the corresponding bit of its second operand. If both bits are 1, the corresponding result bit is set to 1. Otherwise, the corresponding result bit is set to 0.
+ 	
+ 	*  **|** : The bitwise inclusive OR operator compares each bit of its first operand with the corresponding bit of its second operand. If one of the two bits is 1, the corresponding result bit is set to 1. Otherwise, the corresponding result bit is set to 0.
+
+	* **^** : The bitwise exclusive OR (XOR) operator compares each bit of its first operand with the corresponding bit of its second operand. If one bit is 0 and the other bit is 1, the corresponding result bit is set to 1. Otherwise, the corresponding result bit is set to 0.
+
+Examples:
+
+```
+  11111111 11110000
+& 00000000 01100011
+  _________________
+  00000000 01100000
+```
+
+```
+  11111111 11110000
+| 00000000 01100011
+  _________________
+  11111111 11110011
+```
+
+```
+  11111111 11110000
+^ 00000000 01100011
+  _________________
+  11111111 10010011
+```
+
+# Useful tricks.
+
+## Check if a bit is on (to one)
+
+The bit's value can be obtained by shifting the number to the right x times and then bit AND ( & ) on it:
+
+``` cpp
+bool check(int number, int bit) {
+    return ((number >> bit) & 1);
+}
+```
+
+## Change the state of a bit n to x
+
+``` cpp
+number ^= (-x ^ number) & (1 << n);
+```
+
+## Even or Odd
+
+Function to know if a number is even or odd.
+
+True if number is odd, false for number is even.
+
+``` cpp
+bool check(int number) {
+    return (number & 1);
+}
+```
+
+## Check if an integer is a power of 2
+
+``` cpp
+bool powerTwo = n && !(n & (n - 1));
+```
+
+## Brian Kernighan's algorithm.
```

Comment:
I would call this section "Clear the most-right set bit".

And then later in the section say something like: "This can be used to count the set bits of a number. This algorithm is also known as Brian Kernighan's algorithm."